packages/blocks/src/api/registration.js

```diff
@@ -25,12 +25,313 @@ import { isValidIcon, normalizeIconObject } from './utils';
 import { DEPRECATED_ENTRY_KEYS } from './constants';
 
 /**
- * Render behavior of a block type icon; one of a Dashicon slug, an element,
- * or a component.
+ * The `edit` function describes the structure of your block in the context of
+ * the editor. This represents what the editor will render when the block is
+ * used.
  *
- * @typedef {(string|WPElement|WPComponent)} WPBlockTypeIconRender
+ * @callback FnEditBlock
```

Comment:
I would agree in theory, unfortunately I couldn't get JSDoc to understand using any other syntax for documenting functions and their arguments in a detailed manner. I tried `@function` and just adding `@param` tags to the typedefs but the parser was not satisfied AFAICT. Perhaps it's just a limitation we accept, use a generic `Function` and just explain the parameters and return type in the comment block?

Reply:
> I would agree in theory, unfortunately I couldn't get JSDoc to understand using any other syntax for documenting functions and their arguments in a detailed manner.

In a side project, I've had luck with the following form:

```js
/**
 * @typedef {function(SLAuth): Promise<SLStream[]>} SLProviderGetStreams
 */
```

The TypeScript checking verifies the parameter:

![image](https://user-images.githubusercontent.com/1779930/68497921-9d09d680-0223-11ea-9991-103799d9f2cb.png)

...and has Intellisense for the result:

![image](https://user-images.githubusercontent.com/1779930/68497941-abf08900-0223-11ea-87b8-a328d0eb20f1.png)

Reply:
Anything between the brackets in a `typedef` is parsed identically to how typescript `type` definitions are parsed...

So if you have a function that looks like this...

```js
function foo(someString) {
  return Promise.resolve(someString);
}
```

You could make a typedef like this...

```js
/**
 * @typedef {(someString: string) => Promise<string>} FooFunc
 */
```

Or you could reference it inline like this...

```js
/**
 * Function that takes foo as a parameter.
 * @prop {(someString: string) => Promise<string>} fooFunc
 */
function takeFooFunc(fooFunc) { /* ... */ }

Reply:
That's good to know about TS compatibility, thanks! I did get arrow function notation to work with VS Code at least but not seeing a single example anywhere in the JSDoc spec made me nervous about committing to that approach 😅

Reply:
Thanks @dsifford 

@chancestrickland : If it's compatible with (a) TypeScript, (b) `eslint-plugin-jsdoc`, and (c) our documentation generator†, I think it should be fine. I like in the alternate forms @dsifford presents that it gives an option to name the arguments (in my prior comment, it would show as `arg0`).

Reply:
@chancestrickland Yep, I guess I should have mentioned that the way I describe is not strictly aligned with the rules of jsdoc proper. But the editor features and code hints and all that good stuff that are parsed from these type definitions are done so using Microsoft's [tsdoc](https://github.com/microsoft/tsdoc) standards, which is an extension of jsdoc.

So, I suppose it should also be mentioned that doing it this was would _probably_ not pair well with a documentation generator that uses the vanilla jsdoc parser.